File: frontend/src/app/api/models/Membership.ts
import {BaseEntity} from './base/BaseEntity';
import {UserAccount} from './UserAccount';
import {DevTeam} from './DevTeam';
import {MemberType} from './enums/MemberType';

export class Membership extends BaseEntity<Membership> {
  memberType: MemberType;
  userAccount: UserAccount;
  devTeam: DevTeam;

  constructor(memberType: MemberType=MemberType.DEVELOPER, userAccount: UserAccount=null) {
    super();
    this.memberType = memberType;
    this.userAccount = userAccount;
  }

  static isKanbanMaster(m: Membership): boolean {
    return m != null && (
      m.memberType == MemberType.KANBAN_MASTER ||
      m.memberType == MemberType.DEVELOPER_AND_KANBAN_MASTER
    );
  }

  static isProductOwner(m: Membership): boolean {
    return m != null && (
      m.memberType == MemberType.PRODUCT_OWNER ||
      m.memberType == MemberType.DEVELOPER_AND_PRODUCT_OWNER
    );
  }

  static isDeveloper(m: Membership): boolean {
    return m != null &&(
      m.memberType == MemberType.DEVELOPER ||
      m.memberType == MemberType.DEVELOPER_AND_PRODUCT_OWNER ||
      m.memberType == MemberType.DEVELOPER_AND_KANBAN_MASTER
    );
  }

}

File: backend/server/jpa/src/main/java/si/fri/smrpo/kis/server/jpa/entities/Membership.java
package si.fri.smrpo.kis.server.jpa.entities;


import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.voodoodyne.jackson.jsog.JSOGGenerator;
import si.fri.smrpo.kis.server.jpa.entities.base.UUIDEntity;
import si.fri.smrpo.kis.server.jpa.enums.MemberType;

import javax.persistence.*;

@Entity
@Table(name="membership")
@Cacheable
@JsonIdentityInfo(generator=JSOGGenerator.class)
public class Membership extends UUIDEntity<Membership> {

    @Enumerated(EnumType.STRING)
    private MemberType memberType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_account_id", nullable = false)
    private UserAccount userAccount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "dev_team_id", nullable = false)
    private DevTeam devTeam;


    public UserAccount getUserAccount() {
        return userAccount;
    }

    public void setUserAccount(UserAccount userAccount) {
        this.userAccount = userAccount;
    }

    public DevTeam getDevTeam() {
        return devTeam;
    }

    public void setDevTeam(DevTeam devTeam) {
        this.devTeam = devTeam;
    }

    public MemberType getMemberType() {
        return memberType;
    }

    public void setMemberType(MemberType memberType) {
        this.memberType = memberType;
    }

    @JsonIgnore
    public boolean isKanbanMaster() {
        return memberType == MemberType.KANBAN_MASTER || memberType == MemberType.DEVELOPER_AND_KANBAN_MASTER;
    }

    @JsonIgnore
    public boolean isDeveloper() {
        return memberType == MemberType.DEVELOPER || memberType == MemberType.DEVELOPER_AND_KANBAN_MASTER || memberType == MemberType.DEVELOPER_AND_PRODUCT_OWNER;
    }

    @JsonIgnore
    public boolean isProductOwner() {
        return memberType == MemberType.PRODUCT_OWNER || memberType == MemberType.DEVELOPER_AND_PRODUCT_OWNER;
    }

}
